### ISSUE TITLE
regexp_match produces unstable cache keys when used with flags parameter

### ISSUE THREAD

**Issue by lonvia** [sqlalchemy/sqlalchemy#10042]
### Describe the bug

When the regexp_match function is used with the flags parameter (e.g. `regexp_flags('*', flags='i')`) then the resulting construct receives a different cache key for SQL compilation caching every time the function is called. The result is that any SQL containing the function pollutes the compilation cache without ever being reused.

### SQLAlchemy Version in Use

2.0.17

### DBAPI (i.e. the database driver)

psycopg

### Database Vendor and Major Version

Postgresql 15

### Python Version

3.11

### Operating system

Linux

### To Reproduce

```python
from sqlalchemy import column

def make_regexp(text):
    return column('a').regexp_match(text, flags='i')

sql1 = make_regexp('foo')
sql2 = make_regexp('bar')
print(sql1._generate_cache_key())
print(sql2._generate_cache_key())
```

### Error

The output of the test program:
```
CacheKey(key=(
  '0', 
  <class 'sqlalchemy.sql.elements.BinaryExpression'>, 
  'left', 
  (
    '1', 
    <class 'sqlalchemy.sql.elements.ColumnClause'>, 
    'name', 
    'a', 
    'type', 
    (
      <class 'sqlalchemy.sql.sqltypes.NullType'>, 
    ),
  ),
  'right', 
  (
    '2', 
    <class 'sqlalchemy.sql.elements.BindParameter'>, 
    (
      <class 'sqlalchemy.sql.sqltypes.String'>, 
    ),
    '3', 
    False, 
  ),
  'operator', 
  <function regexp_match_op at 0x7f6709d619e0>, 
  'modifiers', 
  (
    (
      'flags', 
      <BindParameter object at 0x7f670af9fa10>, 
    ),
  ),
  'type', 
  (
    <class 'sqlalchemy.sql.sqltypes.Boolean'>, 
    (
      'create_constraint', 
      False, 
    ),
  ),
),)
CacheKey(key=(
  '0', 
  <class 'sqlalchemy.sql.elements.BinaryExpression'>, 
  'left', 
  (
    '1', 
    <class 'sqlalchemy.sql.elements.ColumnClause'>, 
    'name', 
    'a', 
    'type', 
    (
      <class 'sqlalchemy.sql.sqltypes.NullType'>, 
    ),
  ),
  'right', 
  (
    '2', 
    <class 'sqlalchemy.sql.elements.BindParameter'>, 
    (
      <class 'sqlalchemy.sql.sqltypes.String'>, 
    ),
    '3', 
    False, 
  ),
  'operator', 
  <function regexp_match_op at 0x7f6709d619e0>, 
  'modifiers', 
  (
    (
      'flags', 
      <BindParameter object at 0x7f67099e1910>, 
    ),
  ),
  'type', 
  (
    <class 'sqlalchemy.sql.sqltypes.Boolean'>, 
    (
      'create_constraint', 
      False, 
    ),
  ),
),)
``` 

### Additional context

The interesting part of the output above is the cache key part for the flags parameter of regexp_match:

```
  <function regexp_match_op at 0x7f6709d619e0>, 
  'modifiers', 
  (
    (
      'flags', 
      <BindParameter object at 0x7f67099e1910>, 
    ),
  ),
```

As you can see, it uses the BindParameter object itself instead of its more stable cache key. This results in a different key for every call. I would even argue that a BindParameter object is wrong here, at least for the Postgresql dialect where `flags='i'` results in a different operator being used.

**Comment by CaselIT** [comment]
Hi

Thanks for reporting. I think this may be a global problem of the function actually.
Let's wait for mike here.

Did you test also on 1.4?

**Comment by lonvia** [comment]
Yes, same behavior for 1.4.

**Comment by zzzeek** [comment]
it can't be for functions globally because we have a crapload of tests.      Side note it looks like I forgot to add a `__ne__` method to CacheKey() which makes quick testing of this quite annoying, so we should add that also here, but a working test is:

```py
from sqlalchemy import func


x = func.foo_bar(1, 2, 3)
y = func.foo_bar(4, 5, 6)
z = func.foo_bar_bat(1, 2, 3)


assert x._generate_cache_key().key == y._generate_cache_key().key

assert x._generate_cache_key().key != z._generate_cache_key().key

assert not (x._generate_cache_key().key == z._generate_cache_key().key)


```

**Comment by zzzeek** [comment]
regexp_match is an operator, not a function, so this is not related to functions anyway

**Comment by zzzeek** [comment]
so the BinaryElement caches the "flags" as "modifiers" and "modifiers" is not expected to be SQL expressions, it's expected to be constants that are directly cacheable.    at the moment I'm leaning towards keeping that going and not using bound parameters for these flags, but I might reverse course on that.

the other example of "modifiers" are the "escapes" used by LIKE and similar opts.  these are present in the modifiers dictionary directly as strings.      I think we should go with that approach here which requires changes to the compiler, etc.

we might want to add a coercion here that ensures no quote characters are present

**Comment by zzzeek** [comment]
OK we dont need coercsion, we render them as safestirngs in the compiler, great

**Comment by zzzeek** [comment]
ok change of plans again, the "replacement" expression is also placed in the "modifiers".   OK.   I...might want to change that.   I think we need dedicated construct, "modifiers" is not supposed to be used this way

**Comment by CaselIT** [comment]
> ok change of plans again, the "replacement" expression is also placed in the "modifiers". OK. I...might want to change that. I think we need dedicated construct, "modifiers" is not supposed to be used this way

dedicated construct seems to be the best option here

**Comment by zzzeek** [comment]
OK, I have 99% of this working without any new construct.

however.   we have tests with PostgresqL sending a column into "flags".  What's that about?  can we just not support that ?  who would ever do that?

**Comment by CaselIT** [comment]
I don't remember if it was requested or not. pg does support anything everywhere, it just need to have the correct type

I think support for it can be removed and if there are requests it can be re-visited
(should error in that case though)

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**remove use of SQL expressions in "modifiers" for regexp** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4758

**Comment by zzzeek** [comment]
what are the chances we think a third party dialect in 1.4 is using these regexp constructs?  because i have proposed changing their structure to work the way "between" works

**Comment by CaselIT** [comment]
> what are the chances we think a third party dialect in 1.4 is using these regexp constructs? because i have proposed changing their structure to work the way "between" works

haven't checked them but according to github here are 350 matches of visit_regexp_match_op_binary
https://github.com/search?q=visit_regexp_match_op_binary++path%3A*.py+language%3APython&type=Code&ref=advsearch&l=Python&l=

**Comment by zzzeek** [comment]
yes that search is useless because thousands of projects vendor SQLAlchemy in their github repos

**Comment by zzzeek** [comment]
I'd like to go forward and change this to be how I think it should be and if we get regressions from 3rd parties we can help them at that point.   This is at least a way that we get alerted that 3rd parties are even implementing their own regex operators (which I'd wager nobody is right now)

**Comment by zzzeek** [comment]
but i need info on why PG is putting a column expression into "modifiers" and if that is a real use case (looks like not)

**Comment by CaselIT** [comment]
Can you point to the change that added this feature?

**Comment by CaselIT** [comment]
It's likely that it's just a test thing I added.
I can't think of a real use case here

**Comment by zzzeek** [comment]
OK.   just going to go w it....

**Comment by zzzeek** [comment]
also this is not just the flags but the replacement expression also

i know that if I just made "modifiers" cached on SQL expressions that would also fix this, but that was not the intended use of this dictionary....

**Comment by CaselIT** [comment]
More of an use case may be made for replacement compared to flags, but I'm not sure how likely this would be still.

Are there no other cases where there are similar modifiers for operators that may benefit from that feature?

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**remove use of SQL expressions in "modifiers" for regexp** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4759

**Comment by zzzeek** [comment]
> More of an use case may be made for replacement compared to flags, but I'm not sure how likely this would be still.
> 
> Are there no other cases where there are similar modifiers for operators that may benefit from that feature?

the "replacement" part is now moved out of "modifiers" and into the "right" side of the expression; we have a precedent for binary exprs of the form "A operator B secondary C", which is a `BinaryExpression` with a `ClauseList` on the right side.  `between_op` has done this forever.  so that part is handled.

"modifiers" is supposed to be short constants that modify the operator itself.
